evennia/contrib/puzzles.py

```diff
@@ -0,0 +1,533 @@
+
+Puzzles are created from existing objects. The given
+objects are introspected to create prototypes for the
+puzzle parts and results. These prototypes become the
+puzzle recipe. (See PuzzleRecipeObject and @puzzle
+command). Once the recipe is created, all parts and result
+can be disposed (i.e. destroyed).
+
+At a later time, a Builder or a Script can arm the puzzle
+and spawn all puzzle parts (PuzzlePartObject) in their
+respective locations (See @armpuzzle).
+
+A regular player can collect the puzzle parts and combine
+them (See use command). If player has specified
+all pieces, the puzzle is considered solved and all
+its puzzle parts are destroyed while the puzzle results
+are spawened on their corresponding location.
+
+Installation:
+
+Add the PuzzleSystemCmdSet to all players.
+Alternatively:
+
+    @py self.cmdset.add('evennia.contrib.puzzles.PuzzleSystemCmdSet')
+
+"""
+
+import itertools
+from random import choice
+from django.conf import settings
+from evennia import create_object
+from evennia import CmdSet
+from evennia import DefaultObject
+from evennia import DefaultCharacter
+from evennia import DefaultRoom
+from evennia.commands.default.muxcommand import MuxCommand
+from evennia.utils.utils import inherits_from
+from evennia.utils import search, utils, logger
+from evennia.utils.spawner import spawn
+
+# Tag used by puzzles
+_PUZZLES_TAG_CATEGORY = 'puzzles'
+_PUZZLES_TAG_RECIPE = 'puzzle_recipe'
+# puzzle part and puzzle result
+_PUZZLES_TAG_MEMBER = 'puzzle_member'
+
+# ----------- UTILITY FUNCTIONS ------------
+
+def proto_def(obj, with_tags=True):
+    """
+    Basic properties needed to spawn
+    and compare recipe with candidate part
+    """
+    protodef = {
+        # FIXME: Don't we need to honor ALL properties? locks, perms, etc.
+        'key': obj.key,
+        'typeclass': 'evennia.contrib.puzzles.PuzzlePartObject',  # FIXME: what if obj is another typeclass
+        'desc': obj.db.desc,
+        'location': obj.location,
+        # FIXME: Can tags be INVISIBLE? We don't want player to know an object belongs to a puzzle
+        'tags': [(_PUZZLES_TAG_MEMBER, _PUZZLES_TAG_CATEGORY)],
+    }
+    if not with_tags:
+        del(protodef['tags'])
+    return protodef
+
+# ------------------------------------------
+
+class PuzzlePartObject(DefaultObject):
+    """
+    Puzzle Part, typically used by @armpuzzle command
+    """
+
+    def mark_as_puzzle_member(self, puzzle_name):
+        """
+        Marks this object as a member of puzzle named
+        puzzle_name
+        """
+        # FIXME: if multiple puzzles have the same
+        # puzzle_name, their ingredients may be
+        # combined but leave other parts orphan
+        # Similarly, if a puzzle_name were changed,
+        # its parts will become orphan
+        # Perhaps we should use #dbref but that will
+        # force specific parts to be combined
+        self.db.puzzle_name = puzzle_name
+
+
+class PuzzleRecipeObject(DefaultObject):
```

Comment:
Hi @Griatch, I tried replacing `DefaultObject` with `DefaultScript` but got stuck because `evennia.utils.search.search_script()` can only search by `obj.key` and by `obj.dbref`. 

In this case, I need to search by `attribute_name`.

I'm not familiar with Django query syntax but I certainly wouldn't mind "adding" such functionality to a new helper function in `utils.search` *except* for the 0.8 upcoming release.

Being unable to search by `attribute_name` is a roadblock for me.

Suggestions?

Reply:
@Henddher This functionality is available for all typeclassed entities through a manager method, although not through the `search_script` wrapper. Use it like this: 

```python
from evennia import ScriptDB
scripts = ScriptDB.objects.get_by_attribute(key="foo", category="bar", value=None)
```
where `key`, `category` and `value` are all optional and are ways to find scripts only with that Attribute or attribute-value attached to them. The definition is here: https://github.com/evennia/evennia/blob/master/evennia/typeclasses/managers.py#L96

Reply:
Thanks @Griatch - that worked. 

I `git grep`ed and found a thin wrapper of it in `utils.search`. 

https://github.com/evennia/evennia/blob/a53957096f33ce991a52921fe9a26e1984367ede/evennia/utils/search.py#L208-L209

Perhaps that one's "better" for abstraction purposes?